src/main/java/com/aws/greengrass/deployment/IotJobsHelper.java

```diff
@@ -268,17 +275,45 @@ private static void unwrapExecutionException(ExecutionException e)
     }
 
     @Override
-    @SuppressFBWarnings
     public void postInject() {
+        deviceConfiguration.onAnyChange((what, node) -> {
+            if (node != null && what.equals(WhatHappened.childChanged) && relevantNodeChanged(node)) {
+                try {
+                    connectToIotJobs(deviceConfiguration);
+                } catch (DeviceConfigurationException e) {
+                    logger.atWarn().log("Device not configured to talk to AWS Iot cloud. "
+                            + "IOT job deployment is offline: {}", e.getMessage());
+                    return;
+                }
+            }
+        });
+
         try {
-            // Not using isDeviceConfiguredToTalkToCloud() in order to provide the detailed error message to user
-            deviceConfiguration.validate();
+            connectToIotJobs(deviceConfiguration);
         } catch (DeviceConfigurationException e) {
             logger.atWarn().log("Device not configured to talk to AWS Iot cloud. IOT job deployment is offline: {}",
                     e.getMessage());
             return;
         }
+    }
 
+    private boolean relevantNodeChanged(Node node) {
+        // List of configuration nodes that we need to reconfigure for if they change
+        return node.childOf(DEVICE_PARAM_THING_NAME) || node.childOf(DEVICE_PARAM_IOT_DATA_ENDPOINT)
+                || node.childOf(DEVICE_PARAM_PRIVATE_KEY_PATH)
+                || node.childOf(DEVICE_PARAM_CERTIFICATE_FILE_PATH) || node.childOf(DEVICE_PARAM_ROOT_CA_PATH)
+                || node.childOf(DEVICE_PARAM_AWS_REGION);
```

Comment:
these aren't relevant. MQTT client reconfigures itself already. The _only_ thing which should trigger this is when the thing name changes because that actually affects the name of the topic which you're subscribing to.

Reply:
These become relevant when device is transitioning from non-provisioned mode to provisioning. In non-provisioning mode, it may have thingName configured but not other params, and when other params get configured the subscriptions need to be made since those do not exist in the mqtt client. However, we need to re-subscribe only when thing name changes.